File: src/commands/vm.rs
use anyhow::{Result, bail};
use clap::Subcommand;

use crate::client::UnraidClient;
use crate::graphql::get_vms::GetVmsVmsDomains as VmDomain;
use crate::graphql::get_vms::VmState;
use crate::graphql::{ForceStopVm, GetVms, PauseVm, RebootVm, ResetVm, ResumeVm, StartVm, StopVm};

#[derive(Debug, Subcommand)]
pub enum VmCommands {
    /// List virtual machines
    #[command(visible_alias = "ls")]
    List {
        /// Show all VMs (default: only running)
        #[arg(short, long)]
        all: bool,
    },
    /// Start a virtual machine
    Start {
        /// VM name
        name: String,
    },
    /// Stop a virtual machine (graceful shutdown)
    Stop {
        /// VM name
        name: String,
    },
    /// Force stop a virtual machine
    ForceStop {
        /// VM name
        name: String,
    },
    /// Pause a virtual machine
    Pause {
        /// VM name
        name: String,
    },
    /// Resume a paused virtual machine
    Resume {
        /// VM name
        name: String,
    },
    /// Reboot a virtual machine
    Reboot {
        /// VM name
        name: String,
    },
    /// Reset a virtual machine (hard reboot)
    Reset {
        /// VM name
        name: String,
    },
}

pub async fn handle_vm_command(cmd: VmCommands, client: &UnraidClient) -> Result<()> {
    match cmd {
        VmCommands::List { all } => list_vms(client, all).await,
        VmCommands::Start { name } => start_vm(client, &name).await,
        VmCommands::Stop { name } => stop_vm(client, &name).await,
        VmCommands::ForceStop { name } => force_stop_vm(client, &name).await,
        VmCommands::Pause { name } => pause_vm(client, &name).await,
        VmCommands::Resume { name } => resume_vm(client, &name).await,
        VmCommands::Reboot { name } => reboot_vm(client, &name).await,
        VmCommands::Reset { name } => reset_vm(client, &name).await,
    }
}

async fn resolve_vm_id(client: &UnraidClient, name: &str) -> Result<String> {
    let response = client
        .execute::<GetVms>(crate::graphql::get_vms::Variables {})
        .await
        .map_err(map_vms_unavailable)?;

    let domains = response.vms.domains.unwrap_or_default();
    find_vm_id(&domains, name)
}

fn find_vm_id(domains: &[VmDomain], name: &str) -> Result<String> {
    let name_lower = name.to_lowercase();

    for domain in domains {
        if let Some(ref vm_name) = domain.name {
            if vm_name.to_lowercase() == name_lower {
                return Ok(domain.id.clone());
            }
        }
    }

    bail!("VM '{name}' not found. Use 'vm list --all' to see available VMs.");
}

fn map_vms_unavailable(err: anyhow::Error) -> anyhow::Error {
    let msg = err.to_string();
    if msg.contains("VMs are not available") {
        anyhow::anyhow!("VMs are not available on this server. Is the VM service enabled?")
    } else {
        err
    }
}

async fn start_vm(client: &UnraidClient, name: &str) -> Result<()> {
    let id = resolve_vm_id(client, name).await?;

    println!("Starting VM '{name}'...");
    client
        .execute::<StartVm>(crate::graphql::start_vm::Variables { id })
        .await?;

    println!("VM '{name}' start command sent.");
    Ok(())
}

async fn stop_vm(client: &UnraidClient, name: &str) -> Result<()> {
    let id = resolve_vm_id(client, name).await?;

    println!("Stopping VM '{name}'...");
    client
        .execute::<StopVm>(crate::graphql::stop_vm::Variables { id })
        .await?;

    println!("VM '{name}' stop command sent.");
    Ok(())
}

async fn force_stop_vm(client: &UnraidClient, name: &str) -> Result<()> {
    let id = resolve_vm_id(client, name).await?;

    println!("Force stopping VM '{name}'...");
    client
        .execute::<ForceStopVm>(crate::graphql::force_stop_vm::Variables { id })
        .await?;

    println!("VM '{name}' force stop command sent.");
    Ok(())
}

async fn pause_vm(client: &UnraidClient, name: &str) -> Result<()> {
    let id = resolve_vm_id(client, name).await?;

    println!("Pausing VM '{name}'...");
    client
        .execute::<PauseVm>(crate::graphql::pause_vm::Variables { id })
        .await?;

    println!("VM '{name}' pause command sent.");
    Ok(())
}

async fn resume_vm(client: &UnraidClient, name: &str) -> Result<()> {
    let id = resolve_vm_id(client, name).await?;

    println!("Resuming VM '{name}'...");
    client
        .execute::<ResumeVm>(crate::graphql::resume_vm::Variables { id })
        .await?;

    println!("VM '{name}' resume command sent.");
    Ok(())
}

async fn reboot_vm(client: &UnraidClient, name: &str) -> Result<()> {
    let id = resolve_vm_id(client, name).await?;

    println!("Rebooting VM '{name}'...");
    client
        .execute::<RebootVm>(crate::graphql::reboot_vm::Variables { id })
        .await?;

    println!("VM '{name}' reboot command sent.");
    Ok(())
}

async fn reset_vm(client: &UnraidClient, name: &str) -> Result<()> {
    let id = resolve_vm_id(client, name).await?;

    println!("Resetting VM '{name}'...");
    client
        .execute::<ResetVm>(crate::graphql::reset_vm::Variables { id })
        .await?;

    println!("VM '{name}' reset command sent.");
    Ok(())
}

async fn list_vms(client: &UnraidClient, show_all: bool) -> Result<()> {
    let response = client
        .execute::<GetVms>(crate::graphql::get_vms::Variables {})
        .await
        .map_err(map_vms_unavailable)?;

    let domains = response.vms.domains.unwrap_or_default();
    let filtered = filter_by_state(domains, show_all);

    if filtered.is_empty() {
        if show_all {
            println!("No VMs found.");
        } else {
            println!("No running VMs found. Use --all to show all VMs.");
        }
        return Ok(());
    }

    println!("{:<30} {:<12}", "NAME", "STATE");
    println!("{}", "-".repeat(42));

    for vm in filtered {
        let name = vm.name.as_deref().unwrap_or("unnamed");
        let state = format_vm_state(&vm.state);

        println!("{:<30} {:<12}", truncate(name, 29), state);
    }

    Ok(())
}

const fn format_vm_state(state: &VmState) -> &str {
    match state {
        VmState::RUNNING => "running",
        VmState::PAUSED => "paused",
        VmState::SHUTDOWN => "shutdown",
        VmState::SHUTOFF => "shutoff",
        VmState::IDLE => "idle",
        VmState::CRASHED => "crashed",
        VmState::PMSUSPENDED => "suspended",
        VmState::NOSTATE => "no state",
        VmState::Other(_) => "unknown",
    }
}

fn filter_by_state(domains: Vec<VmDomain>, show_all: bool) -> Vec<VmDomain> {
    if show_all {
        domains
    } else {
        domains
            .into_iter()
            .filter(|d| d.state == VmState::RUNNING)
            .collect()
    }
}

fn truncate(s: &str, max_len: usize) -> String {
    if s.len() <= max_len {
        s.to_string()
    } else {
        format!("{}...", &s[..max_len - 3])
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample_vm(id: &str, name: &str, state: VmState) -> VmDomain {
        VmDomain {
            id: id.to_string(),
            name: Some(name.to_string()),
            state,
        }
    }

    fn sample_vms() -> Vec<VmDomain> {
        vec![
            sample_vm("vm-1", "Windows 11", VmState::RUNNING),
            sample_vm("vm-2", "Ubuntu Server", VmState::RUNNING),
            sample_vm("vm-3", "macOS", VmState::SHUTOFF),
            sample_vm("vm-4", "Debian", VmState::PAUSED),
        ]
    }

    // find_vm_id tests

    #[test]
    fn find_vm_id_resolves_matching_names() {
        let vms = sample_vms();
        let cases = [
            ("Windows 11", Ok("vm-1")),
            ("windows 11", Ok("vm-1")),    // case-insensitive
            ("UBUNTU SERVER", Ok("vm-2")), // case-insensitive
            ("nonexistent", Err("not found")),
        ];

        for (input, expected) in cases {
            let result = find_vm_id(&vms, input);
            match expected {
                Ok(id) => assert_eq!(result.unwrap(), id, "find_vm_id({input:?})"),
                Err(msg) => assert!(
                    result.unwrap_err().to_string().contains(msg),
                    "find_vm_id({input:?}) should contain '{msg}'"
                ),
            }
        }
    }

    #[test]
    fn find_vm_id_returns_error_for_empty_list() {
        let err = find_vm_id(&[], "anything").unwrap_err();
        assert!(err.to_string().contains("not found"));
    }

    #[test]
    fn find_vm_id_skips_vms_with_no_name() {
        let vms = vec![VmDomain {
            id: "vm-noname".to_string(),
            name: None,
            state: VmState::RUNNING,
        }];
        let err = find_vm_id(&vms, "something").unwrap_err();
        assert!(err.to_string().contains("not found"));
    }

    // filter_by_state tests

    #[test]
    fn filter_by_state_returns_all_when_show_all_is_true() {
        let vms = sample_vms();
        let filtered = filter_by_state(vms, true);
        assert_eq!(filtered.len(), 4);
    }

    #[test]
    fn filter_by_state_returns_only_running_when_show_all_is_false() {
        let vms = sample_vms();
        let filtered = filter_by_state(vms, false);
        assert_eq!(filtered.len(), 2);
        assert!(filtered.iter().all(|d| d.state == VmState::RUNNING));
    }

    #[test]
    fn filter_by_state_returns_empty_when_no_running_vms() {
        let vms = vec![
            sample_vm("vm-1", "a", VmState::SHUTOFF),
            sample_vm("vm-2", "b", VmState::PAUSED),
        ];
        let filtered = filter_by_state(vms, false);
        assert!(filtered.is_empty());
    }

    // format_vm_state tests

    #[test]
    fn format_vm_state_maps_all_known_states() {
        let cases: Vec<(VmState, &str)> = vec![
            (VmState::RUNNING, "running"),
            (VmState::PAUSED, "paused"),
            (VmState::SHUTDOWN, "shutdown"),
            (VmState::SHUTOFF, "shutoff"),
            (VmState::IDLE, "idle"),
            (VmState::CRASHED, "crashed"),
            (VmState::PMSUSPENDED, "suspended"),
            (VmState::NOSTATE, "no state"),
            (VmState::Other("CUSTOM".to_string()), "unknown"),
        ];

        for (state, expected) in &cases {
            assert_eq!(
                format_vm_state(state),
                *expected,
                "format_vm_state({state:?})"
            );
        }
    }

    // map_vms_unavailable tests

    #[test]
    fn map_vms_unavailable_rewrites_known_error() {
        let cases = [
            (
                "GraphQL errors: Failed to retrieve VM domains: VMs are not available",
                "VM service enabled",
            ),
            ("connection refused", "connection refused"),
        ];

        for (input, expected_substring) in cases {
            let mapped = map_vms_unavailable(anyhow::anyhow!(input.to_string()));
            assert!(
                mapped.to_string().contains(expected_substring),
                "map_vms_unavailable({input:?}) should contain '{expected_substring}'"
            );
        }
    }

    // truncate tests

    #[test]
    fn truncate_handles_various_lengths() {
        let cases = [
            ("hello", 10, "hello"),
            ("hello", 5, "hello"),
            ("hello world", 8, "hello..."),
            ("", 5, ""),
        ];

        for (input, max_len, expected) in cases {
            assert_eq!(
                truncate(input, max_len),
                expected,
                "truncate({input:?}, {max_len})"
            );
        }
    }
}

File: internal/commands/vm.go
package commands

import (
	"context"
	"errors"
	"fmt"
	"log"
	"strings"

	"github.com/Khan/genqlient/graphql"
	"github.com/spf13/cobra"

	"github.com/medzin/unraid-cli/internal/client"
)

type vmDomain = client.GetVmsVmsDomainsVmDomain

type vmMutationFunc func(ctx context.Context, c graphql.Client, id string) error

func newVmCmd(preRun func(*cobra.Command, []string) error) *cobra.Command {
	cmd := &cobra.Command{
		Use:   "vm",
		Short: "Virtual machine management",
	}

	cmd.AddCommand(
		newVmListCmd(),
		newVmStartCmd(),
		newVmStopCmd(),
		newVmForceStopCmd(),
		newVmPauseCmd(),
		newVmResumeCmd(),
		newVmRebootCmd(),
		newVmResetCmd(),
	)
	for _, sub := range cmd.Commands() {
		sub.PreRunE = preRun
	}

	return cmd
}

func newVmListCmd() *cobra.Command {
	var all bool

	cmd := &cobra.Command{
		Use:     "list",
		Aliases: []string{"ls"},
		Short:   "List virtual machines",
		Args:    cobra.NoArgs,
		RunE: func(cmd *cobra.Command, _ []string) error {
			return listVms(cmd.Context(), getClient(cmd.Context()), all)
		},
	}

	cmd.Flags().BoolVarP(&all, "all", "a", false, "show all VMs (default: only running)")
	return cmd
}

func newVmStartCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "start <name>",
		Short: "Start a virtual machine",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return vmAction(cmd.Context(), getClient(cmd.Context()), args[0], "Starting", "start", func(ctx context.Context, c graphql.Client, id string) error {
				_, err := client.StartVm(ctx, c, id)
				return err
			})
		},
	}
}

func newVmStopCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "stop <name>",
		Short: "Stop a virtual machine (graceful shutdown)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return vmAction(cmd.Context(), getClient(cmd.Context()), args[0], "Stopping", "stop", func(ctx context.Context, c graphql.Client, id string) error {
				_, err := client.StopVm(ctx, c, id)
				return err
			})
		},
	}
}

func newVmForceStopCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "force-stop <name>",
		Short: "Force stop a virtual machine",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return vmAction(cmd.Context(), getClient(cmd.Context()), args[0], "Force stopping", "force stop", func(ctx context.Context, c graphql.Client, id string) error {
				_, err := client.ForceStopVm(ctx, c, id)
				return err
			})
		},
	}
}

func newVmPauseCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "pause <name>",
		Short: "Pause a virtual machine",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return vmAction(cmd.Context(), getClient(cmd.Context()), args[0], "Pausing", "pause", func(ctx context.Context, c graphql.Client, id string) error {
				_, err := client.PauseVm(ctx, c, id)
				return err
			})
		},
	}
}

func newVmResumeCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "resume <name>",
		Short: "Resume a paused virtual machine",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return vmAction(cmd.Context(), getClient(cmd.Context()), args[0], "Resuming", "resume", func(ctx context.Context, c graphql.Client, id string) error {
				_, err := client.ResumeVm(ctx, c, id)
				return err
			})
		},
	}
}

func newVmRebootCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "reboot <name>",
		Short: "Reboot a virtual machine",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return vmAction(cmd.Context(), getClient(cmd.Context()), args[0], "Rebooting", "reboot", func(ctx context.Context, c graphql.Client, id string) error {
				_, err := client.RebootVm(ctx, c, id)
				return err
			})
		},
	}
}

func newVmResetCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "reset <name>",
		Short: "Reset a virtual machine (hard reboot)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return vmAction(cmd.Context(), getClient(cmd.Context()), args[0], "Resetting", "reset", func(ctx context.Context, c graphql.Client, id string) error {
				_, err := client.ResetVm(ctx, c, id)
				return err
			})
		},
	}
}

func vmAction(ctx context.Context, c graphql.Client, name, presentVerb, pastVerb string, mutation vmMutationFunc) error {
	id, err := resolveVmID(ctx, c, name)
	if err != nil {
		return err
	}

	log.Printf("%s VM '%s'...", presentVerb, name)

	if err := mutation(ctx, c, id); err != nil {
		return err
	}

	return printAction(ctx, fmt.Sprintf("VM '%s' %s command sent.", name, pastVerb))
}

func resolveVmID(ctx context.Context, c graphql.Client, name string) (string, error) {
	resp, err := client.GetVms(ctx, c)
	if err != nil {
		return "", mapVmsUnavailable(err)
	}
	return findVmID(resp.Vms.Domains, name)
}

func findVmID(domains []vmDomain, name string) (string, error) {
	nameLower := strings.ToLower(name)

	for _, domain := range domains {
		if domain.Name != nil && strings.ToLower(*domain.Name) == nameLower {
			return domain.Id, nil
		}
	}

	return "", fmt.Errorf("VM '%s' not found. Use 'vm list --all' to see available VMs", name)
}

func mapVmsUnavailable(err error) error {
	if strings.Contains(err.Error(), "VMs are not available") {
		return errors.New("VMs are not available on this server. Is the VM service enabled?")
	}
	return err
}

func listVms(ctx context.Context, c graphql.Client, showAll bool) error {
	resp, err := client.GetVms(ctx, c)
	if err != nil {
		return mapVmsUnavailable(err)
	}

	domains := filterVmsByState(resp.Vms.Domains, showAll)

	if domains == nil {
		domains = []vmDomain{}
	}

	return render(ctx, domains, func() error {
		w := getOutputWriter(ctx)
		if len(domains) == 0 {
			if showAll {
				_, err := fmt.Fprintln(w, "No VMs found.")
				return err
			}
			_, err := fmt.Fprintln(w, "No running VMs found. Use --all to show all VMs.")
			return err
		}

		if _, err := fmt.Fprintf(w, "%-30s %-12s\n", "NAME", "STATE"); err != nil {
			return err
		}
		if _, err := fmt.Fprintln(w, strings.Repeat("-", 42)); err != nil {
			return err
		}

		for _, vm := range domains {
			name := "unnamed"
			if vm.Name != nil {
				name = *vm.Name
			}
			if _, err := fmt.Fprintf(w, "%-30s %-12s\n", truncate(name, 29), formatVmState(vm.State)); err != nil {
				return err
			}
		}

		return nil
	})
}

func filterVmsByState(domains []vmDomain, showAll bool) []vmDomain {
	if showAll {
		return domains
	}
	var filtered []vmDomain
	for _, d := range domains {
		if d.State == client.VmStateRunning {
			filtered = append(filtered, d)
		}
	}
	return filtered
}

func formatVmState(state client.VmState) string {
	switch state {
	case client.VmStateRunning:
		return "running"
	case client.VmStatePaused:
		return "paused"
	case client.VmStateShutdown:
		return "shutdown"
	case client.VmStateShutoff:
		return "shutoff"
	case client.VmStateIdle:
		return "idle"
	case client.VmStateCrashed:
		return "crashed"
	case client.VmStatePmsuspended:
		return "suspended"
	case client.VmStateNostate:
		return "no state"
	default:
		return "unknown"
	}
}
